### PR TITLE
Fix race with CTS disposing

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/aspnet/AspNetCore-Internal/issues/2293

`DisposeAsync` will dispose the CTS and then set it to null, however we could still try to cancel it between the dispose and the null resulting in ODE.

Couldn't think of a great way to do this without locking, but it shouldn't be a contested lock so it's fine.

### CODE REVIEW

**Review by mikaelm12 — APPROVED** [review]
Let's see if this fixes the ODE

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs
Comment (halter73) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs]: If you don't keep _stateLock trough the call to Cancel, this could still ODE as this gets disposed by another thread calling DispoasAsync(). After the _stateLock is released by CancelPreviousPoll(), DispoasAsync() can acquire the lock, see the HttpConnectionStatus still isn't dispose and start disposing the Cancellation CTS before CancelPreviousPoll() ever calls Cancel() because the lock was released.

**Review by halter73 — CHANGES_REQUESTED** [review]
We need to keep the _statLock until the HttpConnectionContext is put into a state where it's impossible to have the Cancellation CTS disposed from underneath us.

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs
Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs]: Whoops, I knew that and meant to null it out here...

Comment (halter73) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs]: Cancellation can never get reinitialized, right?

**Review by halter73 — APPROVED** [review]

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs
Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs]: It can, but it's in `TryActivateLongPollingConnection` which is inside the lock

Comment (halter73) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs]: Make sure `TryActivateLongPollingConnection` cannot recreate the CTS. It doesn't look like TryActivateLongPollingConnection checks whether Status is HttpConnectionStatus.Disposed currently.

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs]: We both didn't line up the curly braces correctly when looking at this :D
If the status != inactive it goes to the `FailActivationUnsynchronized` call which will set the correct response and return

File: src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs
Comment (BrennanConroy) [inline src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs]: I believe this test was flaky before (although no failures happened) it was just a very small race, but now that I added another statemachine the test became more flaky and this is my "solution" to make it consistent.

Before we dispatched 2 polls and assumed that one would start before the other, but they actually race each other once the first one goes async. I couldn't find a good place to inject our own sync points so am waiting on the `TransportTask` to change meaning the request has gotten far enough.

If someone has a better idea please help :)

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs]: Might want to add a maximum wait here...

**Comment by BrennanConroy** [comment]
Ping, I modified this a little since when it was approved.

Had to do some ugly test hackery to try to make the test do what it claimed to be doing.

File: src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs
Comment (halter73) [inline src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs]: Where is this var used?

File: src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs
Comment (halter73) [inline src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs]: Shouldn't the test fail if the while loop ends because the count was exceeded?

**Comment by anurse** [comment]
@BrennanConroy do you want to try and get this in today for preview 2? Assuming not (because of the milestone) I've labeled this "NO MERGE" so we hold off on merging until the branch is open for preview 3 changes.

**Comment by BrennanConroy** [comment]
I don't want to rush it, but it would be nice to fix in 3.1. Fixes most LongPolling tests from being flaky

**Comment by anurse** [comment]
Cool. I believe we stay in M2 approval for Preview 3 so let's bring it up then. Especially if it reduces test flakiness :).

**Review by halter73 — APPROVED** [review]

**Comment by anurse** [comment]
Branches are open again for preview 3 work, but this PR does still need @Pilchie approval before merging.

**Comment by Pilchie** [comment]
Approved for 3.1 Preview 3.  Thanks @BrennanConroy.

**Pull request merged**
